app.js

```diff
@@ -486,7 +475,250 @@ define(function(require) {
+					}
+
+					template.find('.feature-key-value').off().on('change', function(event) {
+						if (event.target.className === 'type') {
+							var valueRowIndex = $(this).find('.type').attr('name').match(numberPattern); //Gets the index value for the key value.
+							featureKeys[valueRowIndex - 1].value = event.target.value;
+						}
+
+						if (event.target.parentNode.dataset.type === 'speed_dial') {
+							if (event.target.className === 'sdName') {
+								speedDial.name = event.target.value;
+							}
+
+							if (event.target.className === 'sdNumber') {
+								var valueRowIndex = $(this).find('.sdNumber').attr('name').match(numberPattern); //Gets the index value for the key value.
+								speedDial.number = monster.util.unformatPhoneNumber(event.target.value);
+
+								speedDial.name === '' ? featureKeys[valueRowIndex - 1].value = speedDial.number + ':' + speedDial.number : featureKeys[valueRowIndex - 1].value = speedDial.name + ':' + speedDial.number;
+							}
+						}
+					});
+					$(this).siblings('.feature-key-value.active').removeClass('active'); //Remove the active status for this item.
+				});
+
+				template.find('#proceed').on('click', function() {
+					self.getDeviceFeatureKeyTotal(newUsers, featureKeys);
+
+					self.csvOnboardingRender();
+				});
+
+				template.find('#cancel').on('click', function() {
+					self.csvOnboardingRender();
+				});
+			});
+		},
+
+		getDeviceFeatureKeyTotal: function(newUsers, featureKeys) {
+			var self = this,
+				appData = {
+					deviceList: [],
+					users: newUsers.data,
+					featureKeys: featureKeys
+				};
+
+			_.each(appData.users, function(user) {
+				var device = {
+					brand: user.device.provision.endpoint_brand,
+					family: user.device.provision.endpoint_family,
+					model: user.device.provision.endpoint_model
+				};
+
+				//Check if the device info is in the list if nopt add it
+				self.getDeviceList(device, appData.deviceList);
+			});
+			//Make the API calls to get the number of feature keys allowed on a device.
+			self.getDeviceItteration(appData);
+		},
+
+		getDeviceItteration: function(appData) {
+			var self = this;
+
+			_.each(appData.deviceList, function(device) {
+				monster.request({
+					resource: 'data.template.feature_keys.iteration',
+					data: {
+						brand: device.brand,
+						family: device.family,
+						model: device.model
+					},
+					success: function(apiData, status) {
+						device.keys = apiData.data.template.feature_keys.iterate;
+						self.updateDeviceKeys(appData, device);
+					}
+				});
+			});
+		},
+
+		getDeviceList: function(device, deviceList) {
+			if (!deviceList.length) {
+				deviceList.push(device);
+			} else {
+				//Verify if the targeted device (deviceInfo) has been added to the device list (deviceModels).
+				//This is used narrow down the API calls to the provisioner to get the max feature keys that device can have.
+				_.each(deviceList, function(listDevice) {
+					var found = false;
```

Comment:
This whole _.each section seems unnecessarily complex.  Seems you could replace it with if (device.brand !== listDevice.brand || device.family !== listDevice.family || device.model !== listDevice.model { deviceList.push(device); }

Reply:
This section has been removed in my latest commit and simplified.